sql/logictest: TestSqlLiteLogic failed: column not in input in buildLookupJoin

[(sql/logictest).TestSqlLiteLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2061248&tab=buildLog) on [master@112ff533bf079fb057aadf5a998d9e1f09e6aa1f](https://github.com/cockroachdb/cockroach/commits/112ff533bf079fb057aadf5a998d9e1f09e6aa1f):

```
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_1.test: 9861 statements/queries
rewrote:
CREATE TABLE tab3 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (pk), FAMILY (col3, col4), FAMILY (col1), FAMILY (col5, col2), FAMILY (col0));


rewrote:
CREATE TABLE tab0 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col1, col5), FAMILY (col2), FAMILY (pk, col4), FAMILY (col3, col0));


--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_26.test: 8466 statements/queries
rewrote:
CREATE TABLE tab2 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col0, col4), FAMILY (col3, col5), FAMILY (pk), FAMILY (col1), FAMILY (col2));


rewrote:
CREATE TABLE tab1 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col1), FAMILY (col3, col4, col0, pk, col2), FAMILY (col5));


--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_24.test: 1118 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/1/slt_good_0.test: 6159 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_25.test: 7560 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_28.test: 9460 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_27.test: 6604 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_0.test: 9377 statements/queries
rewrote:
CREATE TABLE tab4 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col1), FAMILY (col0), FAMILY (pk), FAMILY (col2), FAMILY (col5), FAMILY (col4), FAMILY (col3));


--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_26.test: 8834 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_1.test: 9868 statements/queries
rewrote:
CREATE TABLE tab3 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col5), FAMILY (col1, col3, pk, col0), FAMILY (col4, col2));


--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_24.test: 1570 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_25.test: 7709 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/1/slt_good_0.test: 6162 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_28.test: 9909 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_27.test: 7112 statements/queries
rewrote:
CREATE TABLE tab2 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col0, col3), FAMILY (col5), FAMILY (col4), FAMILY (col1), FAMILY (pk, col2));


--- done: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/10/slt_good_28.test with config 3node-tenant: 10033 tests, 0 failures
rewrote:
CREATE TABLE tab0 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col2, col0, col5, col1), FAMILY (col3), FAMILY (col4), FAMILY (pk));


--- total progress: 21779444 statements/queries
        --- FAIL: TestSqlLiteLogic/3node-tenant/slt_good_24.test (42.74s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestSqlLiteLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSqlLiteLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestSqlLiteLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2062825&tab=buildLog) on [master@456a07cfc1e53b87abc7709052e54efb1450e758](https://github.com/cockroachdb/cockroach/commits/456a07cfc1e53b87abc7709052e54efb1450e758):

```
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 3790 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 8343 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9279 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 6538 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 388 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 3963 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 8462 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 2533 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 3947 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 8534 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9446 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 6633 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 542 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 4155 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 8651 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 2724 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 4094 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 8704 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9614 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 6818 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 721 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 4340 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 8846 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 4269 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 2732 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 8829 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9756 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 6909 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 792 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 4466 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 8936 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 4447 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 2883 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 8954 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9879 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 7071 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 954 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 4648 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 9100 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_3.test: 4634 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 2987 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 9072 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9988 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_5.test: 7268 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test: 1143 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_6.test: 9303 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 4847 statements/queries
--- done: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test with config fakedist-spec-planning: 10122 tests, 0 failures
--- total progress: 18745246 statements/queries
        --- FAIL: TestSqlLiteLogic/fakedist-spec-planning/slt_good_12.test#01 (34.00s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestSqlLiteLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSqlLiteLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestSqlLiteLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2063946&tab=buildLog) on [master@3e0de239121813ea4d47873388a2828a66d9edf7](https://github.com/cockroachdb/cockroach/commits/3e0de239121813ea4d47873388a2828a66d9edf7):

```
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_8.test: 9513 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_2.test: 8306 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_0.test: 8676 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_7.test: 3750 statements/queries
rewrote:
CREATE TABLE tab4 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col0), FAMILY (col3, col1), FAMILY (pk, col4), FAMILY (col5), FAMILY (col2));


rewrote:
CREATE TABLE tab0 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col2, col1, col4, col0), FAMILY (col3, col5, pk));


    TestSqlLiteLogic/3node-tenant/slt_good_12.test#01: logic.go:2298: 
        
        /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10634: SELECT pk FROM tab4 WHERE (col0 IN (SELECT col3 FROM tab4 WHERE (col4 = 495.6))) AND (col3 IS NULL)
        expected success, but found
        (XX000) internal error: column 12 not in input
        relational.go:103: in getNodeColumnOrdinal()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:103: getNodeColumnOrdinal()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:570: func1()
        github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:55: func1()
        github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:167: ForEach()
        github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:55: ForEach()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:568: applySimpleProject()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:1381: buildLookupJoin()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:216: buildRelational()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:584: buildProject()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:197: buildRelational()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:136: build()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:105: Build()
        github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:215: makeOptimizerPlan()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:838: makeExecPlan()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:727: dispatchToExecutionEngine()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:500: execStmtInOpenState()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99: execStmt()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1367: execCmd()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1296: run()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491: ServeConn()
        github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:595: func1()
        runtime/asm_amd64.s:1373: goexit()
        
        NOTE: internal errors may have more details in logs. Use -show-logs.
    TestSqlLiteLogic/3node-tenant/slt_good_12.test#01: logic.go:2053: 
         pq: internal error: column 12 not in input
--- done: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test with config 3node-tenant: 2332 tests, 2 failures
    TestSqlLiteLogic/3node-tenant/slt_good_12.test#01: logic.go:2631: 
        /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10638: error while processing
    TestSqlLiteLogic/3node-tenant/slt_good_12.test#01: logic.go:2631: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10638: too many errors encountered, skipping the rest of the input
        --- FAIL: TestSqlLiteLogic/3node-tenant/slt_good_12.test#01 (16.74s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestSqlLiteLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSqlLiteLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestSqlLiteLogic failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2065165&tab=buildLog) on [master@f61d73fe1e4773372f3800b715143305a3342ee7](https://github.com/cockroachdb/cockroach/commits/f61d73fe1e4773372f3800b715143305a3342ee7):

```
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_4.test: 9255 statements/queries
--- progress: /go/src/github.com/cockroachdb/sqllogictest/test/index/delete/10/slt_good_4.test: 9949 statements/queries
rewrote:
CREATE TABLE tab1 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (pk), FAMILY (col2), FAMILY (col0, col5), FAMILY (col4, col3, col1));


rewrote:
CREATE TABLE tab1 (pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING, FAMILY (col2), FAMILY (col5, col4, col1, col0), FAMILY (col3), FAMILY (pk));


    TestSqlLiteLogic/local-vec-auto/slt_good_12.test#01: logic.go:2298: 
        
        /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10634: SELECT pk FROM tab4 WHERE (col0 IN (SELECT col3 FROM tab4 WHERE (col4 = 495.6))) AND (col3 IS NULL)
        expected success, but found
        (XX000) internal error: column 12 not in input
        relational.go:103: in getNodeColumnOrdinal()
        DETAIL: stack trace:
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:103: getNodeColumnOrdinal()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:570: func1()
        github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:55: func1()
        github.com/cockroachdb/cockroach/pkg/util/fast_int_set.go:167: ForEach()
        github.com/cockroachdb/cockroach/pkg/sql/opt/colset.go:55: ForEach()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:568: applySimpleProject()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:1381: buildLookupJoin()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:216: buildRelational()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:584: buildProject()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational.go:197: buildRelational()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:136: build()
        github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:105: Build()
        github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:215: makeOptimizerPlan()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:838: makeExecPlan()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:727: dispatchToExecutionEngine()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:500: execStmtInOpenState()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99: execStmt()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1367: execCmd()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1296: run()
        github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:491: ServeConn()
        github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:595: func1()
        runtime/asm_amd64.s:1373: goexit()
        
        NOTE: internal errors may have more details in logs. Use -show-logs.
    TestSqlLiteLogic/local-vec-auto/slt_good_12.test#01: logic.go:2053: 
         pq: internal error: column 12 not in input
--- done: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test with config local-vec-auto: 2332 tests, 2 failures
    TestSqlLiteLogic/local-vec-auto/slt_good_12.test#01: logic.go:2631: 
        /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10638: error while processing
    TestSqlLiteLogic/local-vec-auto/slt_good_12.test#01: logic.go:2631: /go/src/github.com/cockroachdb/sqllogictest/test/index/commute/100/slt_good_12.test:10638: too many errors encountered, skipping the rest of the input
        --- FAIL: TestSqlLiteLogic/local-vec-auto/slt_good_12.test#01 (11.73s)
=== RUN   TestSqlLiteLogic/local-vec-auto
    --- FAIL: TestSqlLiteLogic/local-vec-auto (0.00s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestSqlLiteLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSqlLiteLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/lookup_join
+++ b/pkg/sql/logictest/testdata/logic_test/lookup_join
@@ -650,3 +650,14 @@ query T
 SELECT message FROM [SHOW TRACE FOR SESSION] WHERE message LIKE 'Scan /Table/72/1/1/{0-1}, /Table/72/1/1/2/{1-2}'
 ----
 Scan /Table/72/1/1/{0-1}, /Table/72/1/1/2/{1-2}
+
+# Regression test for #50964.
+statement ok
+CREATE TABLE tab4 (
+  pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING,
+  UNIQUE (col3 DESC, col4 DESC)
+)
+
+query I
+SELECT pk FROM tab4 WHERE col0 IN (SELECT col3 FROM tab4 WHERE col4 = 495.6) AND (col3 IS NULL)
+----

--- a/pkg/sql/opt/exec/execbuilder/relational.go
+++ b/pkg/sql/opt/exec/execbuilder/relational.go
@@ -1378,7 +1378,11 @@ func (b *Builder) buildLookupJoin(join *memo.LookupJoinExpr) (execPlan, error) {
 
 	// Apply a post-projection if Cols doesn't contain all input columns.
 	if !inputCols.SubsetOf(join.Cols) {
-		return b.applySimpleProject(res, join.Cols, join.ProvidedPhysical().Ordering)
+		outCols := join.Cols
+		if join.JoinType == opt.SemiJoinOp || join.JoinType == opt.AntiJoinOp {
+			outCols = join.Cols.Intersection(inputCols)
+		}
+		return b.applySimpleProject(res, outCols, join.ProvidedPhysical().Ordering)
 	}
 	return res, nil
 }

--- a/pkg/sql/opt/exec/execbuilder/testdata/lookup_join
+++ b/pkg/sql/opt/exec/execbuilder/testdata/lookup_join
@@ -1470,3 +1470,36 @@ limit                                               ·                      ·
 ·                                                   table                  lineitem@primary
 ·                                                   spans                  FULL SCAN
 ·                                                   filter                 l_receiptdate > l_commitdate
+
+# Regression test for #50964.
+statement ok
+CREATE TABLE tab4 (
+  pk INT8 PRIMARY KEY, col0 INT8, col1 FLOAT8, col2 STRING, col3 INT8, col4 FLOAT8, col5 STRING,
+  UNIQUE (col3 DESC, col4 DESC)
+)
+
+query TTTTT
+EXPLAIN (VERBOSE)
+  SELECT pk FROM tab4 WHERE col0 IN (SELECT col3 FROM tab4 WHERE col4 = 495.6) AND (col3 IS NULL)
+----
+·                          distribution           full                                          ·                                          ·
+·                          vectorized             true                                          ·                                          ·
+render                     ·                      ·                                             (pk)                                       ·
+ │                         render 0               pk                                            ·                                          ·
+ └── lookup-join           ·                      ·                                             ("project_const_col_@13", pk, col0, col3)  ·
+      │                    table                  tab4@tab4_col3_col4_key                       ·                                          ·
+      │                    type                   semi                                          ·                                          ·
+      │                    equality               (col0, project_const_col_@13) = (col3, col4)  ·                                          ·
+      │                    equality cols are key  ·                                             ·                                          ·
+      │                    parallel               ·                                             ·                                          ·
+      └── render           ·                      ·                                             ("project_const_col_@13", pk, col0, col3)  ·
+           │               render 0               495.6                                         ·                                          ·
+           │               render 1               pk                                            ·                                          ·
+           │               render 2               col0                                          ·                                          ·
+           │               render 3               col3                                          ·                                          ·
+           └── index-join  ·                      ·                                             (pk, col0, col3)                           ·
+                │          table                  tab4@primary                                  ·                                          ·
+                │          key columns            pk                                            ·                                          ·
+                └── scan   ·                      ·                                             (pk, col3)                                 ·
+·                          table                  tab4@tab4_col3_col4_key                       ·                                          ·
+·                          spans                  /NULL-                                        ·                                          ·

--- a/pkg/sql/opt/ops/relational.opt
+++ b/pkg/sql/opt/ops/relational.opt
@@ -305,10 +305,16 @@ define LookupJoinPrivate {
     # index columns (or a prefix of them).
     KeyCols ColList
 
-    # Cols is the set of columns produced by the lookup join. This set can
-    # contain columns from the input and columns from the index. Any columns not
-    # in the input are retrieved from the index. Cols may not contain some or
-    # all of the KeyCols, if they are not output columns for the join.
+    # Cols is the union between the set of input columns that are returned by
+    # the lookup join and the set of lookup columns retrieved through lookup.
+    #
+    # For inner/left join, this is the set of columns produced by the LookupJoin
+    # operator.
+    # For semi/anti-join, this contains the semi-join output columns (from the
+    # input) plus the columns that we need to look up for the ON condition.
+    #
+    # Cols may not contain some or all of the KeyCols, if they are not output
+    # columns for the join.
     #
     # TODO(radu): this effectively allows an arbitrary projection; it should be
     # just a LookupCols set indicating which columns we should add from the

